ls1intum/Artemis
# `Quiz exercises`: Fix edge cases at the end of the submission period

### Checklist
#### General


#### Server

### Motivation and Context



Currently, if the quiz submission is not allowed anymore (i.e. now > quiz start time + quiz duration + 6 seconds of grace period) but quiz is not yet ended by the instructor (this is possible only for batched & individual quiz), then if the students click View Submission, they cannot see the answers that they have submitted.


**Batched**
https://user-images.githubusercontent.com/10885503/220938198-4a206cb6-888a-4c3a-a6c6-0476cce59ca9.mov

**Individual**
https://user-images.githubusercontent.com/10885503/220938028-b1a07a9d-0b3a-4711-9ae5-dc6752bbb0a6.mov

### Description


The bug happens because if now > quiz start time + quiz duration + 6 seconds of the grace period, the participation state goes back to before state, which means students cannot see the questions before the quiz is started.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course

1. Log in to Artemis
2. Navigate to Course Administration
3. Select the Course
4. Navigate to Exercises
5. Create a Quiz Exercise
6. Select the type that you want to test (Synchronized, Batched, or Individual)
7. Set duration to a shorter time (e.g 30 seconds)
8. Click Save
9. Start the Quiz Exercise (for Synchronized) or add a Batch, then start the Batch (for Batched)
10. Log in as Student
11. Navigate to the Quiz Exercise
12. Submit an answer
13. Before the timer ends, refresh the page or go back and click View Submission (Verify that the answers are displayed)
14. Also do step 13 after the timer ends.

Additional steps:
15. End the Quiz Exercise (for Batched or Individual)
16. Verify that the Student can see the results

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage





### Screenshots


**Batched**
https://user-images.githubusercontent.com/10885503/220937514-8825d667-0f0b-4535-a2b5-e8e31dd95fb6.mov

**Individual**
https://user-images.githubusercontent.com/10885503/220937587-e1fcb8e9-e646-4f61-9e4f-8cce8385a024.mov

**Synchronized**
https://user-images.githubusercontent.com/10885503/220937736-35a0ca23-7549-4a3c-804f-5ad1a2c28562.mov
